File: android_project/folder_type/AlarmManagerExam/app/src/main/java/com/tenisme/alarmmanagerexam/MainActivity.java
package com.tenisme.alarmmanagerexam;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;

// TODO : 안됨..ㅠ 나중에 쓸 거면 로그 찍어보면서 디버깅하든지 하기.

// 안드로이드 내부에서 자체적으로 알람 서비스를 제공함.
// 내부적으로 지정된 시간에 무엇을 하게 할 수 있는 스케줄링 기능을 수행함.
// 통상적인 기상 알람에 사용되는 게 아님.
// 백그라운드에서 어떤 시간을 지정해놓고 그 시간에 무언가를 해야할 때 알람매니저를 수행.
// 알람 매니저(플랫폼이 제공하는 서비스) 는 백그라운드에서 돌아가기 때문에 앱을 종료하고 나서도 타이머가 돌아간다.
// 정확한 시간이 아닐 수 있다.
public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }

    public void showAlarmDialog(View view) {
        // 지정된 시간에 무엇이 동작하게끔 설정하는 가장 간단한 방법은 타임 피커를 "활용"하는 것.
        TimePickerFragment timePickerFragment = new TimePickerFragment();

        // 타임 피커 프레그먼트를 보여줌.
        // getSupportFragmentManager() : 프래그먼트 매니저를 가져옴.
        timePickerFragment.show(getSupportFragmentManager(), "timePicker");
    }
}

File: android_project/folder_type/Room_Exam_Kotlin/app/src/main/java/com/tenisme/room_exam_kotlin/MainActivity.kt
package com.tenisme.room_exam_kotlin

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import com.tenisme.room_exam_kotlin.databinding.ActivityMainBinding
import com.tenisme.room_exam_kotlin.db.entity.Todo
import com.tenisme.room_exam_kotlin.viewmodel.MainViewModel
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // ViewModel + DataBinding 사용 예시
        val binding =
            DataBindingUtil.setContentView<ActivityMainBinding>(this, R.layout.activity_main)
        binding.lifecycleOwner = this // dataBinding 에서 LiveData 를 활용하기 위한 셋팅

        val viewModel = ViewModelProvider(this,
            ViewModelProvider.AndroidViewModelFactory(application))
            .get(MainViewModel::class.java)
        binding.viewModel = viewModel

        viewModel.getAll().observe(this, { binding.editTodo.text = null })

        // 뷰모델 사용 + DataBinding 이 없을 때의 사용 예시
//        viewModel.getAll().observe(this, { txtResult.text = it.toString() })
//
//        btnAdd.setOnClickListener {
//            // 액티비티에서 코루틴 사용하기
//            lifecycleScope.launch(Dispatchers.IO) {
//                // 여기 내부의 코드는 백그라운드에서 돌아간다.
//                viewModel.insert(Todo(editTodo.text.toString())) // 비동기 처리를 해야하는 부분
//            }
//        }

        // 뷰모델이 없을 때 + DataBinding 이 없을 때의 사용 예시
//        val db = Room.databaseBuilder(applicationContext, AppDatabase::class.java, "todo-db")
//            .build()

//        db.todoDao().getAll().observe(this, {
//            txtResult.text = it.toString()
//        })
//
//        btnAdd.setOnClickListener {
//            // 액티비티에서 코루틴 사용하기
//            lifecycleScope.launch(Dispatchers.IO) {
//                // 여기 내부의 코드는 백그라운드에서 돌아간다.
//                db.todoDao().insert(Todo(editTodo.text.toString())) // 비동기 처리를 해야하는 부분
//            }
//        }
    }
}